src/accountsdb/two/Rooted.zig

```diff
@@ -0,0 +1,439 @@
+        \\ PRAGMA page_size = 65536;
+        \\
+        \\CREATE TABLE IF NOT EXISTS entries (
+        \\  address BLOB(32) NOT NULL UNIQUE,
+        \\  lamports INTEGER NOT NULL,
+        \\  data BLOB NOT NULL,
+        \\  owner BLOB(32) NOT NULL,
+        \\  executable INTEGER NOT NULL,
+        \\  rent_epoch INTEGER NOT NULL,
+        \\  last_modified_slot INTEGER NOT NULL
+        \\);
+    ;
+
+    if (sql.sqlite3_exec(db, schema, null, null, null) != OK) {
+        std.debug.print("err  {s}\n", .{sql.sqlite3_errmsg(db)});
+        return error.FailedToCreateTables;
+    }
+
+    return .{
+        .handle = db,
+        .largest_rooted_slot = null,
+    };
+}
+
+pub fn initSnapshot(
+    allocator: std.mem.Allocator,
+    file_path: [:0]const u8,
+    /// Set to directory which contains snapshot account files to pre-load the rooted storage from them.
+    accounts_dir: std.fs.Dir,
+) !Rooted {
+    const zone = tracy.Zone.init(@src(), .{ .name = "Rooted.initSnapshot" });
+    defer zone.deinit();
+
+    const db = blk: {
+        var maybe_db: ?*sql.sqlite3 = null;
+        if (sql.sqlite3_open(file_path.ptr, &maybe_db) != OK)
+            return error.FailedToOpenDb;
+        break :blk maybe_db orelse return error.SqliteDbNull;
+    };
+
+    const db_has_entries = blk: {
+        const query = "SELECT count(*) from entries";
+
+        var stmt: ?*sql.sqlite3_stmt = null;
+        defer if (stmt) |st| std.debug.assert(sql.sqlite3_finalize(st) == OK);
+        const prep_err = sql.sqlite3_prepare_v2(db, query, -1, &stmt, null);
+        if (prep_err != OK) break :blk false; // table does not exist
+
+        const rc = sql.sqlite3_step(stmt);
+        if (rc != ROW) break :blk false; // other err
+
+        break :blk sql.sqlite3_column_int64(stmt, 0) > 0;
+    };
+
+    var self: Rooted = try .init(file_path);
+    if (db_has_entries) {
+        std.debug.print("db has entries, skipping load from snapshot\n", .{});
+    } else {
+        std.debug.print("db is empty -  loading from snapshot!\n", .{});
+        try self.insertFromSnapshot(allocator, accounts_dir);
+    }
+
+    return self;
+}
+
+pub fn deinit(self: *Rooted) void {
+    _ = sql.sqlite3_close(self.handle);
+}
+
+/// Call this before a thread that accesses Rooted closes. Safe to call multiple times.
+pub fn deinitThreadLocals() void {
+    if (put_stmt) |stmt| {
+        _ = sql.sqlite3_finalize(stmt);
+        put_stmt = null;
+    }
+    if (get_stmt) |stmt| {
+        _ = sql.sqlite3_finalize(stmt);
+        get_stmt = null;
+    }
+}
+
+fn accountsHash(self: *Rooted) !sig.core.LtHash {
+    const zone = tracy.Zone.init(@src(), .{ .name = "Rooted.accountsHash" });
+    defer zone.deinit();
+
+    const query =
+        \\SELECT address, owner, data, lamports, executable, rent_epoch, last_modified_slot FROM entries;
+    ;
+
+    var stmt: ?*sql.sqlite3_stmt = undefined;
+    if (sql.sqlite3_prepare_v2(self.handle, query, -1, &stmt, null) != OK)
```

Comment:
`undefined` is probably fine here but makes me uneasy, would prefer erring on the side of distrusting whatever's on the other side of the C ABI boundary and passing initializing to `null`.